aws/resource_aws_ec2_capacity_reservation.go

```diff
@@ -0,0 +1,278 @@
+				Optional: true,
+				ForceNew: true,
+				Default:  ec2.CapacityReservationTenancyDefault,
+				ValidateFunc: validation.StringInSlice([]string{
+					ec2.CapacityReservationTenancyDefault,
+					ec2.CapacityReservationTenancyDedicated,
+				}, false),
+			},
+		},
+	}
+}
+
+func resourceAwsEc2CapacityReservationCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+
+	opts := &ec2.CreateCapacityReservationInput{
+		AvailabilityZone: aws.String(d.Get("availability_zone").(string)),
+		InstanceCount:    aws.Int64(int64(d.Get("instance_count").(int))),
+		InstancePlatform: aws.String(d.Get("instance_platform").(string)),
+		InstanceType:     aws.String(d.Get("instance_type").(string)),
+	}
+
+	if v, ok := d.GetOk("ebs_optimized"); ok {
+		opts.EbsOptimized = aws.Bool(v.(bool))
+	}
+
+	if v, ok := d.GetOk("end_date"); ok {
+		t, err := time.Parse(time.RFC3339, v.(string))
+		if err != nil {
+			return fmt.Errorf("Error parsing capacity reservation end date: %s", err.Error())
+		}
+		opts.EndDate = aws.Time(t)
+	}
+
+	if v, ok := d.GetOk("end_date_type"); ok {
+		opts.EndDateType = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("ephemeral_storage"); ok {
+		opts.EphemeralStorage = aws.Bool(v.(bool))
+	}
+
+	if v, ok := d.GetOk("instance_match_criteria"); ok {
+		opts.InstanceMatchCriteria = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("tenancy"); ok {
+		opts.Tenancy = aws.String(v.(string))
+	}
+
+	restricted := meta.(*AWSClient).IsChinaCloud()
+	if !restricted {
+
+		tagsSpec := make([]*ec2.TagSpecification, 0)
+
+		if v, ok := d.GetOk("tags"); ok {
+			tags := tagsFromMap(v.(map[string]interface{}))
+
+			spec := &ec2.TagSpecification{
+				// There is no constant in the SDK for this resource type
+				ResourceType: aws.String("capacity-reservation"),
+				Tags:         tags,
+			}
+
+			tagsSpec = append(tagsSpec, spec)
+		}
+
+		if len(tagsSpec) > 0 {
+			opts.TagSpecifications = tagsSpec
+		}
+	}
+
+	log.Printf("[DEBUG] Capacity reservation: %s", opts)
+
+	out, err := conn.CreateCapacityReservation(opts)
+	if err != nil {
+		return fmt.Errorf("Error creating capacity reservation: %s", err)
+	}
+	d.SetId(*out.CapacityReservation.CapacityReservationId)
+	return resourceAwsEc2CapacityReservationRead(d, meta)
+}
+
+func resourceAwsEc2CapacityReservationRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+
+	resp, err := conn.DescribeCapacityReservations(&ec2.DescribeCapacityReservationsInput{
+		CapacityReservationIds: []*string{aws.String(d.Id())},
+	})
+
+	if err != nil {
+		// TODO: Check if error is raised if capacity reservation has gone
+		if ec2err, ok := err.(awserr.Error); ok && ec2err.Code() == "InvalidInstanceID.NotFound" {
+			d.SetId("")
+			return nil
+		}
+
+		// Some other error, report it
+		return err
+	}
+
+	// If nothing was found, then return no state
+	if len(resp.CapacityReservations) == 0 {
+		d.SetId("")
```

Comment:
Nit: When removing a resource from the Terraform state, we prefer to provide warning logging, e.g.

```suggestion
		log.Printf("[WARN] EC2 Capacity Reservation (%s) not found, removing from state", d.Id())
		d.SetId("")
```